File: epi_judge_java/epi/SuccessorInTree.java
package epi;

import epi.test_framework.BinaryTreeUtils;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
import epi.test_framework.TimedExecutor;

public class SuccessorInTree {
    static int id = 0;

    public static BinaryTree<Integer> findSuccessor(BinaryTree<Integer> node) {
        id++;
        if (node == null) return null;

        if (node.right == null) {
            while (node.parent != null) {
                if (node == node.parent.left) return node.parent;
                node = node.parent;
            }

            return null;
        } else {
            BinaryTree<Integer> runner = node.right;
            while (runner.left != null) {
                runner = runner.left;
            }

            return runner;
        }
    }

    @EpiTest(testDataFile = "successor_in_tree.tsv")
    public static int findSuccessorWrapper(TimedExecutor executor, BinaryTree<Integer> tree, int nodeIdx)
            throws Exception {
        BinaryTree<Integer> n = BinaryTreeUtils.mustFindNode(tree, nodeIdx);

        BinaryTree<Integer> result = executor.run(() -> findSuccessor(n));

        return result == null ? -1 : result.data;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "SuccessorInTree.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/successor_in_tree.cc
#include "binary_tree_with_parent_prototype.h"
#include "test_framework/binary_tree_utils.h"
#include "test_framework/generic_test.h"

BinaryTreeNode<int> *FindSuccessor(const unique_ptr<BinaryTreeNode<int>> &node) {

    if (node->right != nullptr) {
        auto *runner = node->right.get();
        while (runner->left != nullptr) {
            runner = runner->left.get();
        }

        return runner;
    } else {
        auto *parent = node->parent;
        auto *curr = node.get();

        while (parent != nullptr) {
            if (parent->left.get() == curr) break;

            curr = parent;
            parent = parent->parent;
        }

        return parent;
    }
}

int FindSuccessorWrapper(const unique_ptr<BinaryTreeNode<int>> &tree,
                         int node_idx) {
    auto result = FindSuccessor(MustFindNode(tree, node_idx));
    return result ? result->data : -1;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"tree", "node_idx"};
    return GenericTestMain(args, "successor_in_tree.cc", "successor_in_tree.tsv",
                           &FindSuccessorWrapper, DefaultComparator{},
                           param_names);
}
